# Lecture 1 — Python Basics (Complete, English‑only)

This notebook is an **English translation** of the original lecture.

### How to use this notebook
- Read the explanation cell(s) first, then run the code cells to see the concepts in action.
- `input()` examples are simulated so you can run everything smoothly in Jupyter.


## Why Python is popular
- Interest in Python has grown strongly since around 2011 on communities such as Stack Overflow; it often ranks at or near the top among languages like Java, JavaScript, and C#.
- Python is heavily used for data analysis, data processing, and machine learning/AI by companies such as Google, Meta (Facebook), Uber, Instagram, Netflix, PayPal, and Dropbox.
- It is general‑purpose, easy to read, and supported by a rich standard library and third‑party ecosystem.
- Python’s simple syntax makes it accessible for beginners and powerful for professionals.


## Open source and community
- Python is free and open source.
- The standard library and many external packages provide ready‑to‑use tools for numerous domains.
- A large, active community makes it easy to find answers and examples.


## Interpreted language
C++ and similar languages are typically compiled to a binary (e.g., `.exe`) before execution. Python code is usually executed line by line by an interpreter. Tools exist to package Python applications, but the normal workflow is interpreted.


## Tools for programming in Python
- PyCharm (IDE)
- Visual Studio Code (editor with Python extension)
- Jupyter (in‑browser notebooks)
- Replit (in‑browser coding)

Useful links:
- PyCharm intro (YouTube): https://www.youtube.com/watch?v=56bPIGf4us0
- VS Code + Python: https://code.visualstudio.com/docs/python/python-tutorial
- Try Jupyter online: https://jupyter.org/try
- Replit: https://replit.com/


## Standard Python Modules

A standard Python installation includes a library of standard modules.

A **module** is a file with the `.py` extension that contains a collection of functions, classes, variables, constants, and other Python objects used to solve specific tasks.

The contents of a module can be accessed and used by any other Python program.

You can display a list of the standard modules by running the following command:

```python
help("modules")
```

In [5]:
help('modules')

Total modules found: 901
First 25 modules: ['69e79323355fec38eab6__mypyc', '81d243bd2c585b0f4821__mypyc', 'Cryptodome', 'Cython', 'IPython', 'OpenGL', 'OpenSSL', 'PIL', '__future__', '__hello__', '__phello__', '_aix_support', '_argon2_cffi_bindings', '_asyncio', '_brotli', '_bz2', '_cffi_backend', '_codecs_cn', '_codecs_hk', '_codecs_iso2022', '_codecs_jp', '_codecs_kr', '_codecs_tw', '_collections_abc', '_compat_pickle']


A module is imported using the following command:

```python
import module_name
```

Multiple modules can be imported at the same time:

```python
import module_name_1, module_name_2, ..., module_name_n
```

To call a function from a module, write the module name followed by a dot (`.`) and the function name:

```python
module_name.function_name()
```

In [6]:
import math

print(math.factorial(5))      # 120
x = math.pi
print(math.cos(x))            # -1.0


120
-1.0


In [7]:
import math, random

print(math.degrees(math.asin(1)))  # 90.0
print(random.randint(-5, 5))  # random integer in [-5, 5]

90.0
5


A module can also be imported using a different name (an **alias**):

```python
import module_name as new_module_name
```

In [8]:
import math as m
print(m.pi)
print(m.e)

3.141592653589793
2.718281828459045


The `from` statement allows you to import specific functions from a module:

```python
from module_name import function_name[, ..., function_name_n]
```

After importing the functions this way, you can call them directly without writing the module name:

```python
function_name()
```

In [9]:
from math import sin, pi
print(sin(pi/4))  # 0.7071067811865476

0.7071067811865475


## Programming style (PEP 8)
Most programming languages can be written (or, more accurately, formatted) in different styles. Some styles are easier to read than others. Writing code that is easy for others to understand is a good practice, and using a consistent coding style helps make programs more readable.

Most Python projects follow the style guidelines defined in **PEP 8**, which promote clear and readable code. Some of the most important recommendations are:

- Use **4 spaces** for indentation, and do not use tabs.
- Use blank lines to separate functions, classes, and larger blocks of code.
- Whenever possible, write comments on a separate line.
- Use **docstrings** to document functions, classes, and modules.
- Use spaces around operators and after commas, but not inside parentheses:

```python
a = f(1, 2) + g(3, 4)
```

- Use consistent naming conventions:
  - **Classes:** `CamelCase`
  - **Functions and methods:** `lower_case_with_underscores`
  - Always use `self` as the first parameter of instance methods.

## Variables and data types
### Variable Names

Variable names in Python must follow these rules:

- They may contain letters, numbers, and underscores (`_`).
- They cannot begin with a number.
- They cannot contain spaces.
- Python is **case-sensitive**, so uppercase and lowercase letters are treated as different.
- A variable name cannot be the same as a Python keyword.

You can display the list of Python keywords using the `keyword` module:


In [10]:
import keyword
print(keyword.kwlist)
print(keyword.iskeyword('if'))
print(keyword.iskeyword('iff'))


['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']
True
False


Python is an **object-oriented programming language**, where all elements (numbers, strings, functions, etc.) are treated as objects.

A **variable name** is a reference to an object.

A value is assigned using the `=` symbol:

```python
variable_name = value
```

In [ ]:
a = 100
b = 'one hundred'
print(a, b)

Python allows you to assign the same or different values to multiple variables at the same time:

In [11]:
a = b = c = 10
print(a, b, c)

x, y, z = 1, 2, 'five'
print(x, y, z)

10 10 10
1 2 five


If an assigned value is too long, it can be continued on the next line using the `\` symbol:

In [12]:
line = 'the quick brown fox jumps \
over a lazy dog'
print(line)

the quick brown fox jumps over a lazy dog


If brackets `[ ]`, `{ }`, or `( )` are used, there is no need to use the line continuation symbol `\`:

In [13]:
A = [1, 2, 3,
     4, 5, 6,
     7, 8, 9]
print(A)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


The `#` symbol indicates that everything written after it is a comment.

A comment can consist of only one line. If a comment needs to span multiple lines, the `#` symbol must be written at the beginning of each line.

## Core data types
Python is a **dynamically typed programming language**. This means that variables do not have a fixed type; instead, the type belongs to the value assigned to the variable.

The type of a variable can be determined using the `type` command:

```python
type(variable_name)
```

### Integers

In [15]:
print(2 + 3)
x = 10
print(type(x), x)

5
<class 'int'> 10


### Float


In [16]:
a = 2.1
b = 3 * a
print(b, type(b))
print(-6.3e2)
a = 10.2e-4
print(a, type(a))

6.300000000000001 <class 'float'>
-630.0
0.00102 <class 'float'>


### Complex Numbers

Complex numbers consist of a real part (`.real`) and an imaginary part (`.imag`):

In [17]:
c = 5.1 + 0.1j
print(c.real, c.imag, type(c))
d = 2 - 1.4j
print(c - d)

5.1 0.1 <class 'complex'>
(3.0999999999999996+1.5j)


### Binary Integers (`bin`)

Binary integers are base-2 whole numbers. They start with `0b` or `0B` (zero followed by the letter `b` or `B`).

In [ ]:
print(0b11011)   # 27
print(bin(15))

### Octal Integers (`oct`)

Octal integers are base-8 whole numbers. They start with `0o` or `0O` (zero followed by the letter `o` or `O`).

In [ ]:
print(0o45601)   # 19329
print(oct(2020))

### Hexadecimal Integers (`hex`)

Hexadecimal integers are base-16 whole numbers. They start with `0x` or `0X` (zero followed by the letter `x` or `X`).

In [18]:
print(0x15ab3)   # 88755
print(hex(1027297))

88755
0xface1


### Booleans

In [ ]:
print(1 > 3)
a, b = 2, 3
print(a != b)
boo = (3 > 1)
print(boo, type(boo))

When performing operations with Boolean variables:

```python
True = 1
False = 0
```

In [19]:
print(True + True)
print(False - True)
print(True * False)
try:
    print(True / False)
except ZeroDivisionError as e:
    print('ZeroDivisionError:', e)

2
-1
0
ZeroDivisionError: division by zero


### Strings
Strings are assigned using single or double quotation marks:

In [ ]:
s1 = 'string value'
s2 = "string value"
print(type(s1), type(s2))

s = '"Yes," he said'
print(s)
s = '\"Isn\'t,\" she said'
print(s)
s = '''"Isn't," she said'''
print(s)
multi = '''Hello
World'''
print(multi)


An empty string can be created.

The built-in `len()` function returns the length of a string.

In [ ]:
print(len(''))   # empty string -> 0
print(len(' '))  # space -> 1

If a string contains a single quotation mark, it can be enclosed in double quotation marks, and vice versa. Alternatively, the escape character `\` can be used.

Triple single or double quotation marks can also be used.
```

In [1]:
s = '"Yes," he said'
print(s)
s = '\"Isn\'t,\" she said'
print(s)
s = '''"Isn't," she said'''
print(s)
multi = '''Hello
World'''
print(multi)

"Yes," he said
"Isn't," she said
"Isn't," she said
Hello
World


Triple quotation marks allow you to create a string with the desired format.

In [2]:
multi = '''Hello
World'''
print(multi)

Hello
World


Strings can be joined using the `+` operator and repeated using the `*` operator:

In [ ]:
print('Hello' + ' ' + 'world')
print('Hello ' * 3 + 'world')
print(5 * 'Hello ')

Strings written next to each other are automatically joined:

In [3]:
print('Hello' ' ' 'world')

Hello world


The `str()` function converts non-string data types into strings:

In [4]:
try:
    _ = 'Apollo' + 11
except TypeError as e:
    print('TypeError:', e)
print('Apollo' + str(11))

TypeError: can only concatenate str (not "int") to str
Apollo11


Characters in a string are indexed, and indexes are written inside square brackets. The first character of a string has index `0`.

Example:

| Index (positive) |     0 |     1 |     2 |     3 |     4 |     5 |     6 |     7 |     8 |     9 |    10 |
|------------------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|
| String           |   H   |   e   |   l   |   l   |   o   | space |   w   |   o   |   r   |   l   |   d   |
| Index (negative) |   -11 |  -10  |   -9  |   -8  |   -7  |   -6  |   -5  |   -4  |   -3  |   -2  |   -1  |

The nth element of a string sequence is accessed using its index:

```python
string[n]
```

Negative indexing can also be used:

```python
-n = len(string) - n
```

If the index value is outside the sequence boundaries, an error message is displayed.

In [6]:
string = 'Hello world'
print(string[0])
print(string[-1])
print(string[len(string)-1])
try:
    print(string[11])
except IndexError as e:
    print('IndexError:', e)

H
d
d
IndexError: string index out of range
lowr
dlrow olleH


Operations can be performed on parts of sequences using **slice notation**.

String slices are defined using two indexes separated by a colon:

```python
string[i : j]
```

**Important:** The character with index `j` (`string[j]`) is not included in the slice. The slice ends with the character `string[j-1]`.

Slice indexes can be omitted:
- If the first index is missing, `0` is used.
- If the second index is missing, the length of the string is used.

An index that is larger than the string length is replaced with the string length.

In [7]:
print(string[0:10])
print(string[0:11])
print(string[2:7])
print(string[:5])
print(string[6:])
print(string[-5:]) # 5 last symbols

Hello worl
Hello world
llo w
Hello
world
world


A string is an immutable sequence of characters. This means that individual characters in a string cannot be changed.

However, new strings can be created using methods and functions.
```

In [8]:
s = 'Hello world'
try:
    s[6] = 'W'
except TypeError as e:
    print('TypeError:', e)
s = s[:6] + 'W' + s[7:]
print(s)
s = s[:6] * 3 + s[6:] + '!'
print(s)


TypeError: 'str' object does not support item assignment
Hello World
Hello Hello Hello World!


### Extended Slice Syntax

The extended slice syntax is:

```python
string[i : j : k]
```

Here, `k` specifies the step size of the slice.

In [10]:
print(string[2:9:2])

lowr


If the step is negative, the slice is returned in reverse order.

In [9]:
print(string[::-1])

dlrow olleH


## Data Type conversion
A data type can be converted using the functions `int()`, `float()`, `bool()`, `complex()`, and `str()`.

### `int()`

In [11]:
print(int(10.999))  # float > int
try:
    print(int(10 + 5j)) # complex > int
except TypeError as e:
    print('TypeError:', e)
print(int(True), int(False)) # bool > int
print(int('20')) # str > int
for bad in ['10.5', '0b1111']:
    try:
        print(int(bad))
    except ValueError as e:
        print('ValueError for', bad + ':', e)


10
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'complex'
1 0
20
ValueError for 10.5: invalid literal for int() with base 10: '10.5'
ValueError for 0b1111: invalid literal for int() with base 10: '0b1111'


### `float()`

In [12]:
print(float(100)) #int > float
print(float(0b1111))
print(float(0o1234))
print(float(0xFACE3))
try:
    print(float(10 + 5j)) # complex > float
except TypeError as e:
    print('TypeError:', e)
print(float(True), float(False)) # bool > float
print(float('10')) # str > float
print(float('12.345'))


100.0
15.0
668.0
1027299.0
TypeError: float() argument must be a string or a real number, not 'complex'
1.0 0.0
10.0
12.345


### `complex()`

**Form 1:**

```python
complex(x)
```

In [ ]:
print(complex(10)) # int > complex
print(complex(0b1110))
print(complex(10.5)) # float > complex
print(complex(True), complex(False)) # bool > complex
print(complex('10')) # str > complex
print(complex('10.55'))
for bad in ['0b1010']:
    try:
        print(complex(bad))
    except ValueError as e:
        print('ValueError for', bad + ':', e)

**Form 2:**

```python
complex(x, y)
```

In [13]:
print(complex(10, 5)) # int > complex
print(complex(0b111, 5))
print(complex(10.5, 5.2)) # float > complex
for bad in [('10', '5'), (10, '5')]: # str > complex
    try:
        print(complex(*bad))
    except TypeError as e:
        print('TypeError for', bad, ':', e)

(10+5j)
(7+5j)
(10.5+5.2j)
TypeError for ('10', '5') : complex() can't take second arg if first is a string
TypeError for (10, '5') : complex() second arg can't be a string




```
# This is formatted as code
```

### `bool()`

When converting to the Boolean (`bool`) type, all numbers except `0` are considered `True`, and all strings except empty strings (`''`) are considered `True`.

In [ ]:
for v in [-100, 10.999, 0.0001, 10+5j, 0b1111, 0, 'Hello World', ' ', '']:
    print(repr(v), '->', bool(v))


###`str()`

In [14]:
print(str(10))
print(str(0b11001))
print(str(10.99))
print(str(1 + 0j))
print(str(True), str(False))

10
25
10.99
(1+0j)
True False


## Mutable and Immutable Objects

A **mutable object** is an object that can change its value while keeping the same `id()`.

An **immutable object** is an object with a fixed value.

If we want to change the value of an immutable object, a new object must be created.

### The `id()` Function

Every Python object is stored in a specific memory location.

The `id()` function returns the memory location of a given object. The address is returned as a base-10 number.

This function can be used to check whether objects refer to the same memory location, meaning they refer to the same data in memory.

### The `is` Keyword

`is` is a keyword that compares two objects and returns `True` if both objects have the same memory address.

`is not` returns `False` if both objects have the same memory address.

### Immutable Objects

- Numbers (`int`, `float`, `decimal`, `boolean`, `complex`)
- Strings
- Tuples
- Frozen sets

In [ ]:
a = 10
print('a =', a, ' id(a)=', id(a))
a = 20
print('a =', a, ' id(a)=', id(a))

a = 1000
b = a
c = a
print('id(a)=', id(a))
print('id(b)=', id(b))
print('id(c)=', id(c))
print('a is b ->', a is b)
print('a is c ->', a is c)
print('b is c ->', b is c)

a = 10
b = a
print('a is b ->', a is b)
a = 1000
print('a is b ->', a is b)

x = []
y = []
print('x == y ->', x == y)
print('x is y ->', x is y)
print('x is not y ->', x is not y)


Mutable objects include:

- Lists
- Sets
- Dictionaries

A **list** is a collection of values of any data type, separated by commas and enclosed in square brackets.

In [ ]:
my_list = [10, 20, 30, 40]
print(id(my_list))
my_list[2] = 3000
print(id(my_list))
print(my_list)

list1 = [10, 20, 30, 40]
list2 = list1
print('list1 is list2 ->', list1 is list2)
list1[0] = 100.52
print('list1 is list2 ->', list1 is list2)
print('list1 =', list1)
print('list2 =', list2)

list1 = [10, 20, 30, 40]
list2 = [10, 20, 30, 40]
print('id(list1)=', id(list1))
print('id(list2)=', id(list2))
print('list1 is list2 ->', list1 is list2)

t = (1, 2, 3)
try:
    t[0] = 99
except TypeError as e:
    print('TypeError:', e)

s = {1, 2, 2, 3}
s.add(4)
print('set:', s)
d = {'a': 1, 'b': 2}
d['c'] = 3
print('dict:', d)
fs = frozenset([1, 2, 2, 3])
print('frozenset:', fs)


## Input and Output

### Data Input

Data is entered from the keyboard using the `input()` function.

All data read from the keyboard is treated as strings (`<class 'str'>`).

To avoid calculation errors, the input data type must be converted during input or before performing calculations.

In [15]:
a = (input('First number: '))
b = (input('Second number: '))
print('a + b =', int(a)+int(b))

a = int(input('First number: '))
b = int(input('Second number: '))
print('a + b =', a+b)


First number: 5
Second number: 7
a + b = 12
First number: 2
Second number: 8
a + b = 10


### The `eval()` Function

In [16]:
b = bool(input('Enter True or False: '))
print('b =', b)

Enter True or False: True
b = True


When converting to the Boolean (`bool`) type, all strings except empty strings (`''`) are considered `True`. Therefore, in this case, the `eval()` function is used for type conversion:

In [18]:
b = eval(input(' Enter True or False: '))
print('b =', b)
b = eval(input('Enter Expression: '))
print('b =', b)

 Enter True or False: True
b = True
Enter Expression: 10+3
b = 13


How to read multiple values from the keyboard in one line:

In [19]:
s = (input('Enter 3 numbers: '))# read str() type object
print(s)
L = s.split() # splits values, craetes list() type object
print(L)
L_int = [int(x) for x in L]#str > int
print(L_int)
a, b, c = L_int
print('a =', a)
print('b =', b)
print('c =', c)

Enter 3 numbers: 3 5 4
3 5 4
['3', '5', '4']
[3, 5, 4]
a = 3
b = 5
c = 4


All performed steps are combined into a single line:

In [20]:
a, b, c = [int(x) for x in (input('Enter 3 numbers: ')).split()]
print('a =', a)
print('b =', b)
print('c =', c)

a, b, c = [eval(x) for x in (input('Enter some values: ')).split(',')]
print('a =', a, type(a))
print('b =', b, type(b))
print('c =', c, type(c))

Enter 3 numbers: 3 5 6
a = 3
b = 5
c = 6


### Data Output

In [21]:
print('Hello\nWorld')
print()
print('Hello\tWorld')

Hello
World

Hello	World


The `sep` Attribute

In [ ]:
x, y = 2.5, 19
print(x, y, x*y)
print(x, y, x*y, sep='-----')
print(x, y, x*y, sep='      ')

## The `end` Attribute

In [22]:
x, y = 2.5, 19
print(x, end='*****')
print(y, end='\n\n\n')
print(x*y)

2.5*****19


47.5


## Formatted output
### The String Formatting Method `.format()`

In [ ]:
first_name = 'John'
last_name = 'Smith'
salary = 2500
print('Hello {} {}. Your salary is {}'.format(first_name, last_name, salary))
print('Hello {0} {1}. Your salary is {2}'.format(first_name, last_name, salary))
print('Hello {1} {0}. Your salary is {2}'.format(first_name, last_name, salary))
print('Hello {a} {b}. Your salary is {c}'.format(a=first_name, b=last_name, c=salary))
print('Hello {} {}. Your salary is {}'.format('John', 'Smith', 2500))


### Output Formats

- `:d` – integer (`int`)
- `:f` – floating-point number (`float`)
- `:e` – exponential notation
- `:s` – string (`str`)

In [24]:
n = 458.541315
try:
    print('{:d}'.format(n))  # wrong: n is float
except ValueError as e:
    print('ValueError:', e)
print('{:d}'.format(int(n)))
print('{:f}'.format(n))
print('{0:15f}'.format(n))
print('{0:15e}'.format(n))
print('{0:15.3f}'.format(n))

ValueError: Unknown format code 'd' for object of type 'float'
458
458.541315
     458.541315
   4.585413e+02
        458.541


### Alignment

- `<` – align to the left
- `>` – align to the right
- `^` – align to the center

In [25]:
print('{0:15f}{1}'.format(n, '|'))
print('{0:<15f}{1}'.format(n, '|'))
print('{0:>15f}{1}'.format(n, '|'))
print('{0:^15.3f}{1}'.format(n, '|'))
print('{:s}'.format('hello'))

     458.541315|
458.541315     |
     458.541315|
    458.541    |
hello


## Formatting Using f-strings

This formatting method has been available since Python 3.6.

To format a string, the letter `f` is placed before the string.

In [26]:
x = 356.089775672
print(f'Value of x: {x}')
print(f'{x = }')
width = 10
precision = 4
y = 12.3456789
print(f'Value of y: {y:{width}.{precision}}')


Value of x: 356.089775672
x = 356.089775672
Value of y:      12.35


# Operators

## Arithmetic Operators

| Operator | Meaning | Examples |
|---|---|---|
| `+` | Addition | `1 + 2 → 3`<br>`'abc' + 'defg' → 'abcdefg'` |
| `-` | Subtraction | `5.23 - 10 → -4.77` |
| `*` | Multiplication | `10 * 10 → 100`<br>`'abc' * 2 → 'abcabc'` |
| `/` | Division (result is `float`) | `10 / 2 → 5.0` |
| `//` | Integer division (floor rounding) | `10.2 // 6 → 1.0`<br>`-10.2 // 6 → -2.0` |
| `%` | Remainder of division | `10 % 3 → 1`<br>`3.5 % 2 → 1.5` |
| `divmod(x, y)` | Returns a pair `(x // y, x % y)` | `divmod(10, 3) → (3, 1)` |
| `**` | Exponentiation | `10 ** 3 → 1000` |
| `pow(x, y)` | Exponentiation `xʸ` | `pow(27, 2/3)` |
| `abs(x)` | Absolute value | `abs(-10) → 10` |
| `min(s)` | Smallest element in a sequence | `min('120254854945') → '0'` |
| `max(s)` | Largest element in a sequence | `max('120254854945') → '9'` |

---

## Shortened Assignment Operators

| Operator | Equivalent |
|---|---|
| `x += y` | `x = x + y` |
| `x -= y` | `x = x - y` |
| `x *= y` | `x = x * y` |
| `x /= y` | `x = x / y` |
| `x //= y` | `x = x // y` |
| `x %= y` | `x = x % y` |
| `x **= y` | `x = x ** y` |

---

## Comparison Operators

| Operator | Meaning |
|---|---|
| `==` | Equal |
| `!=` | Not equal |
| `>` | Greater than |
| `>=` | Greater than or equal to |
| `<` | Less than |
| `<=` | Less than or equal to |

---

## Logical Operators

### `not` — Logical NOT

| a | not a |
|---|---|
| True | False |
| False | True |

### `and` — Logical AND

| a | b | a and b |
|---|---|---|
| True | True | True |
| True | False | False |
| False | True | False |
| False | False | False |

### `or` — Logical OR

| a | b | a or b |
|---|---|---|
| True | True | True |
| True | False | True |
| False | True | True |
| False | False | False |

---

## Checking Membership

Example:

```python
eilute = 'Hello world'
```

| Operator | Meaning | Example |
|---|---|---|
| `in` | Returns `True` if a value exists in a sequence | `'H' in eilute → True`<br>`'W' in eilute → False` |
| `not in` | Returns `True` if a value does not exist in a sequence | `'abc' not in eilute → True`<br>`'llo' not in eilute → False` |

## The `math` module — commonly used functions and constants
- `ceil(x)`: smallest integer greater than or equal to `x`
- `copysign(x, y)`: `x` with the sign of `y`
- `fabs(x)`: absolute value as float
- `factorial(x)`: `x!`
- `floor(x)`: largest integer less than or equal to `x`
- `fmod(x, y)`: remainder of `x / y`
- `frexp(x)`: mantissa and exponent of `x` as `(m, e)`
- `fsum(iterable)`: accurate floating‑point sum
- `isfinite(x)`: True if not infinity and not NaN
- `isinf(x)`: True if positive or negative infinity
- `isnan(x)`: True if NaN (Not a Number)
- `ldexp(x, i)`: `x * (2**i)`
- `modf(x)`: fractional and integer parts of `x`
- `trunc(x)`: truncated integer value of `x`
- `exp(x)`, `expm1(x)`: `e**x`, `e**x - 1`
- `log(x[, base])`, `log1p(x)`, `log2(x)`, `log10(x)`
- `pow(x, y)`, `sqrt(x)`
- `acos(x)`, `asin(x)`, `atan(x)`, `atan2(y, x)`
- `cos(x)`, `sin(x)`, `tan(x)`, `hypot(x, y)`
- `degrees(x)`, `radians(x)`
- `acosh(x)`, `asinh(x)`, `atanh(x)`, `cosh(x)`, `sinh(x)`, `tanh(x)`
- `erf(x)`, `erfc(x)`, `gamma(x)`, `lgamma(x)`
- Constants: `pi`, `e`